File: frontend/types/pre-planejamento.ts
// Espelha backend/app/schemas/pre_planejamento.py

export interface Study {
  id: string;
  projectId: string;
  name: string;
  startDate: string;
  durationMonths: number | null;
  createdAt: string;
}

export interface Service {
  id: string;
  name: string;
  color: string;
  orderIndex: number;
  lagDays: number;
}

export interface Floor {
  id: string;
  groupName: string;
  floorName: string;
  orderIndex: number;
}

export interface Cycle {
  id: string;
  serviceId: string;
  floorId: string;
  durationDays: number;
}

export interface Holiday {
  id: string;
  date: string;
  description: string;
  isNational: boolean;
}

export interface Predecessor {
  cycleId: string;
  predecessorId: string;
}

export interface StudyDetail extends Study {
  services: Service[];
  floors: Floor[];
  cycles: Cycle[];
  holidays: Holiday[];
  predecessors: Predecessor[];
}

export interface CreateStudyInput {
  name: string;
  startDate: string;
  durationMonths: number;
}

export interface HolidayInput {
  date: string;
  description: string;
  isNational: boolean;
}

export interface UpdateStudyInput {
  name: string;
  startDate: string;
  holidays: HolidayInput[];
}

export interface ServiceInput {
  id?: string;
  name: string;
  color: string;
  orderIndex: number;
  lagDays: number;
}

export interface FloorInput {
  id?: string;
  groupName: string;
  floorName: string;
  orderIndex: number;
}

export interface CycleInput {
  serviceIndex: number;
  floorIndex: number;
  durationDays: number;
}

export interface SaveCyclesInput {
  services: ServiceInput[];
  floors: FloorInput[];
  cycles: CycleInput[];
}

export interface WbsOverrideInput {
  cycleId: string;
  predecessorIds: string[];
}

export interface SaveWbsOverridesInput {
  overrides: WbsOverrideInput[];
}

File: backend/app/schemas/pre_planejamento.py
"""Schemas de request/response de /api/v1/pre-planejamento."""
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel


class StudyOut(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    start_date: date
    created_at: datetime


class ServiceOut(BaseModel):
    id: UUID
    name: str
    color: str
    order_index: int
    lag_days: int


class FloorOut(BaseModel):
    id: UUID
    group_name: str
    floor_name: str
    order_index: int


class CycleOut(BaseModel):
    id: UUID
    service_id: UUID
    floor_id: UUID
    duration_days: int


class HolidayOut(BaseModel):
    id: UUID
    date: date
    description: str
    is_national: bool


class StudyDetailOut(StudyOut):
    services: list[ServiceOut]
    floors: list[FloorOut]
    cycles: list[CycleOut]
    holidays: list[HolidayOut]


class CreateStudyRequest(BaseModel):
    name: str
    start_date: date


class HolidayIn(BaseModel):
    date: date
    description: str
    is_national: bool = False


class UpdateStudyRequest(BaseModel):
    name: str
    start_date: date
    holidays: list[HolidayIn]


class ServiceIn(BaseModel):
    name: str
    color: str
    order_index: int
    lag_days: int = 0


class FloorIn(BaseModel):
    group_name: str
    floor_name: str
    order_index: int


class CycleIn(BaseModel):
    service_index: int
    floor_index: int
    duration_days: int


class SaveCyclesRequest(BaseModel):
    services: list[ServiceIn]
    floors: list[FloorIn]
    cycles: list[CycleIn]
